les/vflux/server/service.go

```diff
@@ -0,0 +1,116 @@
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package server
+
+import (
+	"net"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/les/utils"
+	"github.com/ethereum/go-ethereum/les/vflux"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+type (
+	// Server serves vflux requests
+	Server struct {
+		limiter         *utils.Limiter
+		services        map[string]*serviceEntry
+		delayPerRequest time.Duration
+	}
+
+	// Service is a service registered at the Server and identified by a string id
+	Service interface {
+		ServiceInfo() (id, desc string)                                      // only called during registration
+		Handle(id enode.ID, address string, name string, data []byte) []byte // never called concurrently
+	}
+
+	serviceEntry struct {
+		id, desc string
+		backend  Service
+	}
+)
+
+// NewServer creates a new Server
+func NewServer(delayPerRequest time.Duration) *Server {
+	return &Server{
+		limiter:         utils.NewLimiter(1000),
+		delayPerRequest: delayPerRequest,
+		services:        make(map[string]*serviceEntry),
+	}
+}
+
+// Register registers a Service
+func (s *Server) Register(b Service) {
+	srv := &serviceEntry{backend: b}
+	srv.id, srv.desc = b.ServiceInfo()
+	if strings.Contains(srv.id, ":") {
+		// srv.id + ":" will be used as a service database prefix
+		log.Error("Service ID contains ':'", "id", srv.id)
+		return
+	}
+	s.services[srv.id] = srv
+}
+
+// Serve serves a vflux request batch
+// Note: requests are served by the Handle functions of the registered services. Serve
+// may be called concurrently but the Handle functions are called sequentially and
+// therefore thread safety is guaranteed.
+func (s *Server) Serve(id enode.ID, address string, requests vflux.Requests) vflux.Replies {
+	reqLen := uint(len(requests))
+	if reqLen == 0 || reqLen > vflux.MaxRequestLength {
+		return nil
+	}
+	// Note: the value parameter will be supplied by the token sale module (total amount paid)
+	ch := <-s.limiter.Add(id, address, 0, reqLen)
+	if ch == nil {
+		return nil
+	}
+	// Note: the following section is protected from concurrency by the limiter
+	results := make(vflux.Replies, len(requests))
+	for i, req := range requests {
+		if service := s.services[req.Service]; service != nil {
```

Comment:
`s.services` is map. We should add lock protection.

Reply:
Although there is only read-read contention

Reply:
Not even that, the limiter avoids all concurrency. Still I added a lock (and a comment) because it costs nothing and now it's safe even if a new service is registered after the handler has been started.